File: Development Tools/PFQuest Conversion/analyze_pfquest_data.py
#!/usr/bin/env python3

import re
import os
from datetime import datetime

def extract_lua_table(content, table_name):
    """Extract a Lua table from content"""
    pattern = rf'{table_name}\s*=\s*\{{'
    start = re.search(pattern, content)
    if not start:
        return {}
    
    # Simple extraction - find matching braces
    brace_count = 0
    start_idx = start.end() - 1
    end_idx = start_idx
    
    for i in range(start_idx, len(content)):
        if content[i] == '{':
            brace_count += 1
        elif content[i] == '}':
            brace_count -= 1
            if brace_count == 0:
                end_idx = i + 1
                break
    
    table_str = content[start_idx:end_idx]
    return parse_lua_table(table_str)

def parse_lua_table(table_str):
    """Parse a Lua table string into a Python dict (simplified)"""
    result = {}
    
    # Extract quest entries [ID] = { ... }
    pattern = r'\[(\d+)\]\s*=\s*(\{[^}]*\}|"[^"]*")'
    matches = re.finditer(pattern, table_str)
    
    for match in matches:
        quest_id = int(match.group(1))
        value = match.group(2)
        
        if value.startswith('"'):
            # String value (placeholder quest)
            result[quest_id] = value.strip('"')
        else:
            # Table value (actual quest data)
            result[quest_id] = value
    
    return result

def load_pfquest_data():
    """Load pfQuest-epoch database files"""
    base_path = "/Users/travisheryford/Library/CloudStorage/Dropbox/WoW Interfaces/epoch/AddOns/pfQuest-epoch/db/"
    
    # Load quest data
    with open(base_path + "quests-epoch.lua", 'r', encoding='utf-8') as f:
        quest_content = f.read()
    
    # Load quest text (English)
    with open(base_path + "enUS/quests-epoch.lua", 'r', encoding='utf-8') as f:
        text_content = f.read()
    
    # Parse quest IDs and basic info
    quests = {}
    quest_pattern = r'\[(\d+)\]\s*=\s*\{([^}]+(?:\{[^}]*\}[^}]*)*)\}'
    
    for match in re.finditer(quest_pattern, quest_content):
        quest_id = int(match.group(1))
        quest_data = match.group(2)
        
        # Skip removed quests
        if '"_"' in quest_data or '= "_"' in quest_content[match.start():match.end()]:
            continue
            
        quests[quest_id] = {
            'raw': quest_data,
            'level': None,
            'start_npcs': [],
            'end_npcs': []
        }
        
        # Extract level
        level_match = re.search(r'\["lvl"\]\s*=\s*(\d+)', quest_data)
        if level_match:
            quests[quest_id]['level'] = int(level_match.group(1))
        
        # Extract start NPCs
        start_match = re.search(r'\["start"\]\s*=\s*\{[^}]*\["U"\]\s*=\s*\{([^}]+)\}', quest_data)
        if start_match:
            npc_ids = re.findall(r'\d+', start_match.group(1))
            quests[quest_id]['start_npcs'] = [int(npc) for npc in npc_ids]
        
        # Extract end NPCs
        end_match = re.search(r'\["end"\]\s*=\s*\{[^}]*\["U"\]\s*=\s*\{([^}]+)\}', quest_data)
        if end_match:
            npc_ids = re.findall(r'\d+', end_match.group(1))
            quests[quest_id]['end_npcs'] = [int(npc) for npc in npc_ids]
    
    # Parse quest text
    texts = {}
    text_pattern = r'\[(\d+)\]\s*=\s*\{([^}]+(?:\[[^\]]*\][^}]*)*)\}'
    
    for match in re.finditer(text_pattern, text_content):
        quest_id = int(match.group(1))
        text_data = match.group(2)
        
        texts[quest_id] = {}
        
        # Extract title
        title_match = re.search(r'\["T"\]\s*=\s*"([^"]+)"', text_data)
        if title_match:
            texts[quest_id]['title'] = title_match.group(1)
        
        # Extract objectives
        obj_match = re.search(r'\["O"\]\s*=\s*"([^"]+)"', text_data)
        if obj_match:
            texts[quest_id]['objectives'] = obj_match.group(1)
    
    return quests, texts

def load_questie_data():
    """Load Questie-Epoch database files"""
    base_path = "/Users/travisheryford/Library/CloudStorage/Dropbox/WoW Interfaces/epoch/AddOns/Questie/Database/Epoch/"
    
    # Load quest data
    with open(base_path + "epochQuestDB.lua", 'r', encoding='utf-8') as f:
        content = f.read()
    
    quests = {}
    # Pattern for Questie quest entries
    pattern = r'\[(\d+)\]\s*=\s*\{([^}]+(?:\{[^}]*\}[^}]*)*)\}'
    
    for match in re.finditer(pattern, content):
        quest_id = int(match.group(1))
        quest_data = match.group(2)
        
        # Extract quest name (first field)
        name_match = re.search(r'^"([^"]+)"', quest_data)
        if name_match:
            quests[quest_id] = {
                'name': name_match.group(1),
                'raw': quest_data
            }
    
    return quests

def main():
    print("Loading pfQuest-epoch data...")
    pf_quests, pf_texts = load_pfquest_data()
    
    print("Loading Questie-Epoch data...")
    questie_quests = load_questie_data()
    
    print(f"\n=== ANALYSIS RESULTS ===")
    print(f"pfQuest-epoch has {len(pf_quests)} quests")
    print(f"Questie-Epoch has {len(questie_quests)} quests")
    
    # Find missing quests
    missing_in_questie = []
    placeholder_names = []
    
    for quest_id in pf_quests:
        if quest_id not in questie_quests:
            if quest_id in pf_texts and 'title' in pf_texts[quest_id]:
                missing_in_questie.append(quest_id)
        elif questie_quests[quest_id]['name'].startswith("[Epoch] Quest"):
            if quest_id in pf_texts and 'title' in pf_texts[quest_id]:
                placeholder_names.append(quest_id)
    
    print(f"Missing in Questie: {len(missing_in_questie)} quests")
    print(f"Placeholder names in Questie: {len(placeholder_names)} quests")
    
    # Generate missing quests file
    print("\n=== MISSING QUESTS (First 30) ===")
    output_path = "/Users/travisheryford/Library/CloudStorage/Dropbox/WoW Interfaces/epoch/AddOns/Questie/"
    
    with open(output_path + "pfquest_missing_quests.txt", 'w') as f:
        f.write(f"-- Missing quests from pfQuest-epoch\n")
        f.write(f"-- Generated on {datetime.now()}\n\n")
        
        count = 0
        for quest_id in sorted(missing_in_questie)[:30]:
            if quest_id in pf_texts:
                title = pf_texts[quest_id].get('title', 'Unknown')
                level = pf_quests[quest_id].get('level', '?')
                objectives = pf_texts[quest_id].get('objectives', '')
                
                print(f"  [{quest_id}] = \"{title}\" (Level {level})")
                
                # Generate Questie format
                start_npcs = pf_quests[quest_id]['start_npcs']
                end_npcs = pf_quests[quest_id]['end_npcs']
                
                f.write(f'[{quest_id}] = {{"{title}",')
                
                if start_npcs:
                    f.write('{{' + ','.join(map(str, start_npcs)) + '}},')
                else:
                    f.write('nil,')
                
                if end_npcs:
                    f.write('{{' + ','.join(map(str, end_npcs)) + '}},')
                else:
                    f.write('nil,')
                
                f.write(f'nil,{level or 1},nil,nil,')
                
                if objectives:
                    f.write('{"' + objectives.replace('"', '\\"') + '"},')
                else:
                    f.write('nil,')
                
                f.write('nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,0,nil,nil,nil,nil,nil,nil},')
                f.write(' -- From pfQuest-epoch\n')
                
                count += 1
    
    # Generate name updates file
    print("\n=== PLACEHOLDER NAME UPDATES (First 30) ===")
    
    with open(output_path + "pfquest_name_updates.txt", 'w') as f:
        f.write(f"-- Quest name updates from pfQuest-epoch\n")
        f.write(f"-- Generated on {datetime.now()}\n\n")
        
        count = 0
        for quest_id in sorted(placeholder_names)[:30]:
            if quest_id in pf_texts:
                current_name = questie_quests[quest_id]['name']
                new_name = pf_texts[quest_id].get('title', 'Unknown')
                
                print(f"  [{quest_id}] = \"{current_name}\" -> \"{new_name}\"")
                
                f.write(f"-- [{quest_id}] = \"{current_name}\" -> \"{new_name}\"\n")
                count += 1
    
    print(f"\n=== FILES GENERATED ===")
    print(f"1. pfquest_missing_quests.txt - {len(missing_in_questie)} missing quests (showing first 30)")
    print(f"2. pfquest_name_updates.txt - {len(placeholder_names)} quests with placeholder names (showing first 30)")

if __name__ == "__main__":
    main()

File: Validation/archive/temp_lua/analyze_pfquest_data.lua
#!/usr/bin/env lua

-- Script to analyze and compare pfQuest-epoch data with Questie-Epoch data
-- and generate conversion suggestions

-- Helper function to safely load Lua files
local function loadLuaFile(filepath)
    local file = io.open(filepath, "r")
    if not file then
        print("Error: Could not open file " .. filepath)
        return nil
    end
    
    local content = file:read("*all")
    file:close()
    
    -- Create sandbox environment
    local env = {
        pfDB = {},
        epochQuestData = {},
        epochNpcData = {},
        QuestieDB = { _epochQuestData = {} },
        QuestieLoader = { ImportModule = function() return {} end }
    }
    
    local chunk, err = loadstring(content)
    if not chunk then
        print("Error loading " .. filepath .. ": " .. err)
        return nil
    end
    
    setfenv(chunk, env)
    local success, err = pcall(chunk)
    if not success then
        print("Error executing " .. filepath .. ": " .. err)
        return nil
    end
    
    return env
end

-- Load pfQuest data
local pfQuestPath = "/Users/travisheryford/Library/CloudStorage/Dropbox/WoW Interfaces/epoch/AddOns/pfQuest-epoch/"
local questiePath = "/Users/travisheryford/Library/CloudStorage/Dropbox/WoW Interfaces/epoch/AddOns/Questie/"

print("Loading pfQuest databases...")
local pfQuestData = loadLuaFile(pfQuestPath .. "db/quests-epoch.lua")
local pfQuestText = loadLuaFile(pfQuestPath .. "db/enUS/quests-epoch.lua")
local pfUnitsData = loadLuaFile(pfQuestPath .. "db/units-epoch.lua")

print("Loading Questie databases...")
local questieQuestData = loadLuaFile(questiePath .. "Database/Epoch/epochQuestDB.lua")
local questieNpcData = loadLuaFile(questiePath .. "Database/Epoch/epochNpcDB.lua")

if not pfQuestData or not pfQuestText or not questieQuestData then
    print("Failed to load required databases")
    return
end

-- Extract the actual data
local pfQuests = pfQuestData.pfDB["quests"]["data-epoch"] or {}
local pfTexts = pfQuestText.pfDB["quests"]["enUS-epoch"] or {}
local pfUnits = pfUnitsData and pfUnitsData.pfDB["units"]["data-epoch"] or {}
local questieQuests = questieQuestData.epochQuestData or {}
local questieNpcs = questieNpcData and questieNpcData.epochNpcData or {}

-- Analyze data
local pfQuestCount = 0
local questieQuestCount = 0
local missingInQuestie = {}
local missingNames = {}

-- Count pfQuest quests
for questId, questData in pairs(pfQuests) do
    if type(questData) == "table" then
        pfQuestCount = pfQuestCount + 1
        
        -- Check if quest exists in Questie
        if not questieQuests[questId] then
            missingInQuestie[questId] = {
                data = questData,
                text = pfTexts[questId]
            }
        elseif questieQuests[questId] and questieQuests[questId][1] == "[Epoch] Quest " .. questId then
            -- Quest exists but has placeholder name
            missingNames[questId] = {
                data = questData,
                text = pfTexts[questId],
                questieEntry = questieQuests[questId]
            }
        end
    end
end

-- Count Questie quests
for questId, questData in pairs(questieQuests) do
    if type(questData) == "table" then
        questieQuestCount = questieQuestCount + 1
    end
end

print("\n=== ANALYSIS RESULTS ===")
print(string.format("pfQuest-epoch has %d quests", pfQuestCount))
print(string.format("Questie-Epoch has %d quests", questieQuestCount))
print(string.format("Missing in Questie: %d quests", table.getn(missingInQuestie)))
print(string.format("Placeholder names in Questie: %d quests", table.getn(missingNames)))

-- Helper function to convert faction
local function convertFaction(raceFlag)
    if not raceFlag then return 3 end -- Both factions
    if raceFlag == 77 then return 2 end -- Alliance (1+4+8+64 = Human+NE+Gnome+Draenei)
    if raceFlag == 178 then return 1 end -- Horde (2+16+32+128 = Orc+Tauren+Troll+BE)
    return 3 -- Default to both
end

-- Generate conversion suggestions for missing quests
print("\n=== MISSING QUESTS TO ADD ===")
local outputFile = io.open(questiePath .. "pfquest_missing_quests.lua", "w")
outputFile:write("-- Missing quests from pfQuest-epoch that should be added to Questie\n")
outputFile:write("-- Generated on " .. os.date() .. "\n\n")
outputFile:write("local missingQuests = {\n")

local count = 0
for questId, questInfo in pairs(missingInQuestie) do
    if count < 20 then -- Limit output for readability
        local text = questInfo.text or {}
        local data = questInfo.data
        
        if text.T then -- Only process if we have a title
            print(string.format("  [%d] = \"%s\" (Level %s)", 
                questId, text.T or "Unknown", data.lvl or "?"))
            
            -- Generate Questie format entry
            local startNpcs = data.start and data.start.U or nil
            local endNpcs = data["end"] and data["end"].U or nil
            local faction = convertFaction(data.race)
            
            outputFile:write(string.format('  [%d] = {"%s",', questId, text.T or "Unknown"))
            
            -- Start NPCs
            if startNpcs then
                outputFile:write("{{" .. table.concat(startNpcs, ",") .. "}},")
            else
                outputFile:write("nil,")
            end
            
            -- End NPCs
            if endNpcs then
                outputFile:write("{{" .. table.concat(endNpcs, ",") .. "}},")
            else
                outputFile:write("nil,")
            end
            
            -- Other fields
            outputFile:write(string.format('nil,%s,nil,%d,', data.lvl or 1, faction))
            
            -- Objectives
            if text.O then
                outputFile:write(string.format('{"%s"},', text.O:gsub('"', '\\"')))
            else
                outputFile:write("nil,")
            end
            
            -- Fill remaining fields with nil
            outputFile:write("nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,nil,0,nil,nil,nil,nil,nil,nil},")
            outputFile:write(" -- From pfQuest-epoch\n")
            
            count = count + 1
        end
    end
end

outputFile:write("}\n\n")
outputFile:write("-- Total missing: " .. table.getn(missingInQuestie) .. " quests\n")
outputFile:close()

-- Generate update suggestions for placeholder names
print("\n=== QUESTS WITH PLACEHOLDER NAMES TO UPDATE ===")
local updateFile = io.open(questiePath .. "pfquest_name_updates.lua", "w")
updateFile:write("-- Quests with placeholder names that can be updated from pfQuest-epoch\n")
updateFile:write("-- Generated on " .. os.date() .. "\n\n")
updateFile:write("local nameUpdates = {\n")

count = 0
for questId, questInfo in pairs(missingNames) do
    if count < 20 then
        local text = questInfo.text or {}
        if text.T then
            print(string.format("  [%d] = \"%s\" -> \"%s\"", 
                questId, questInfo.questieEntry[1], text.T))
            
            updateFile:write(string.format('  [%d] = {current = "%s", updated = "%s"},\n',
                questId, questInfo.questieEntry[1], text.T))
            count = count + 1
        end
    end
end

updateFile:write("}\n\n")
updateFile:write("-- Total with placeholder names: " .. table.getn(missingNames) .. " quests\n")
updateFile:close()

print("\n=== FILES GENERATED ===")
print("1. pfquest_missing_quests.lua - Quests to add to Questie")
print("2. pfquest_name_updates.lua - Quest names to update in Questie")
print("\nReview these files to see conversion suggestions!")